In [81]:
from math import factorial as fl
from math import exp
import numpy as np

In [82]:
def C(k,n):
    return int(fl(n) / (fl(k) * fl(n-k)))

Задача 1
Контрольная работа состоит из пяти вопросов. На каждый вопрос приведено четыре варианта ответа, один из которых правильный. Составьте закон распределения числа правильных ответов при простом угадывании. Найдите M(X), D(X).

Число испытаний n = 5, вероятность вытащить правильный билет 1/4:

$$P(X = k) = C_n^k p^k (1 - p)^{n - k}$$

У нас могут быть следующие события: X - правильный ответ на: 0, 1, 2, 3, 4, 5 вопросов.

In [83]:
x_values = np.arange(0, 6)
print(x_values)

[0 1 2 3 4 5]


Тогда вероятности каждого события будут следующие:

In [84]:
n = 5
p = 0.25
x_probabilities = np.array([])

In [85]:
for k in x_values:
    x_probabilities = np.append(x_probabilities, (C(k,n)*p**k*(1-p)**(n-k)))

In [86]:
x_probabilities

array([0.23730469, 0.39550781, 0.26367188, 0.08789062, 0.01464844,
       0.00097656])

In [87]:
m = x_values.dot(x_probabilities)
m

1.25

In [88]:
y_values = x_values - m
print(y_values)

[-1.25 -0.25  0.75  1.75  2.75  3.75]


In [89]:
z_values = y_values ** 2
print(z_values)

[ 1.5625  0.0625  0.5625  3.0625  7.5625 14.0625]


In [90]:
d = z_values.dot(x_probabilities)
print(d)

0.9375


Ответ: математическое ожидание = 1.25, дисперсия = 0.9375

Задача 2
Вероятность попадания в цель при одном выстреле равна 0.01. Какова вероятность того, что число попаданий при 200 выстрелах будет не менее 5 и не более 10?

In [95]:
p = 0.01

In [96]:
n = 200

$$P(X = k) = \dfrac{a^k e ^{-a}}{k!}.$$

In [99]:
a = n * p
a

2.0

In [100]:
P_total = 0
for k in range(5, 11):
    P_k = (a**k * exp(-a))/fl(k)
    P_total = P_total + P_k
P_total

0.052644709119342686

Задача 3
В условиях предыдущей задачи сколько выстрелов нужно сделать чтобы быть уверенным с вероятностью 0.9, что хотя бы 10 раз будет совершено попадание?
Подсказка. Как посчитать P(X >= k) для величины X, имеющей распределение Пуассона? Здесь предстоит немножко покодить.

$$P(X = k) = \dfrac{a^k e ^{-a}}{k!}.$$

In [132]:
p = 0.01
for n in range(100,2000):
    a = n * p
    P_total = 0
    for k in range(0, 10):
        P_k = (a**k * exp(-a))/fl(k)
        P_total = P_total + P_k
    if (1 - P_total) >= 0.9:
        break
print(n)
print(P_total)

1421
0.09982392770158856
